### PR TITLE
feat: BGD review fixes

### PR DESCRIPTION
**Pull request by defijesus**
## Motivation
BGD review fixes

### CODE REVIEW

File: src/AFeiToDaiSwapper.sol [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+        IERC20(DAI).approve(address(AAVE_LENDING_POOL), type(uint256).max);
+    }
+
+    function swapAllAvailable() external {
+        uint256 redeemAmount = IERC20(FEI).balanceOf(address(AAVE_LENDING_POOL));
+        
+        uint256 aFeiReserveBalance = IERC20(A_FEI).balanceOf(AAVE_MAINNET_RESERVE_FACTOR);
+        if (aFeiReserveBalance < redeemAmount) {
+            redeemAmount = aFeiReserveBalance;
+        }
```
Comment (0xrajath) [inline src/AFeiToDaiSwapper.sol]: Just curious as to the need for this check? In what situation do you foresee the `aaFeiReserveBalance < redeemAmount` ?

Comment (defijesus) [reply src/AFeiToDaiSwapper.sol]: As we redeem more and more chunks of aFEI from the reserve, at one point we might have more liquidity in the pool than we have in the reserves. At that point, instead of withdrawing max liquidity, we simply withdraw max balance available. This edge case will be the last time the function will be called and allows the reserve to be completely depleted.

File: src/test/ValidateRedeemFei.sol [hunk 2]
```diff
@@ -23,24 +37,31 @@ contract ValidationRedeemFei is Test {
...
-        feiPayload = new RedeemFei();
+        swapper = new AFeiToDaiSwapper();
+        feiPayload = new RedeemFei(address(swapper));
     }
 
     function testProposalPostPayload() public {
 
-        uint256 aFeiAmount = 300_000e18;
-        uint256 daiBalanceBefore = IERC20(DAI).balanceOf(AAVE_MAINNET_RESERVE_FACTOR);
-        uint256 aFeiBalanceBefore = IERC20(A_FEI).balanceOf(AAVE_MAINNET_RESERVE_FACTOR);
```
Comment (0xrajath) [inline src/test/ValidateRedeemFei.sol:43]: Can we have the check for aFeiBalanceBefore and aFeiBalanceAfter as well to show that it's all been used up?

Comment (defijesus) [reply src/test/ValidateRedeemFei.sol:43]: It shouldnt be all used up until the last time the function is called, maybe I'm misunderstanding

File: src/test/ValidateRedeemFei.sol [hunk 3]
```diff
@@ -68,12 +89,15 @@ contract ValidationRedeemFei is Test {
...
+        uint256 aDaiReserveBalanceAfter = IERC20(A_DAI).balanceOf(AAVE_MAINNET_RESERVE_FACTOR);
+        emit log_named_uint("Amount of FEI in LendingPool", aFeiPoolBalance);
+        emit log_named_uint("Max DAI Redeem from PSM after 3bps fee -----------", minBalance);
+        emit log_named_uint("PSM expected output amount from FEI in LendingPool", psmAmountOut);
+        emit log_named_uint("aDAI balance before ----------------", aDaiReserveBalanceBefore);
+        emit log_named_uint("aDAI balance after -----------------", aDaiReserveBalanceAfter);
+        emit log_named_uint("Difference between expected and real", aDaiReserveBalanceAfter - (aDaiReserveBalanceBefore + minBalance));
 
-        uint256 minBalance = aFeiAmount - (aFeiAmount * 3 / 10_000);
+        assertGe(aDaiReserveBalanceAfter, aDaiReserveBalanceBefore + minBalance);
```
Comment (0xrajath) [inline src/test/ValidateRedeemFei.sol:100]: Can you comment the reason for using `assertGe` here. I.e that AAVE Mainnet Reserve Factor gets some additional aTokens minted to it while redeeming. 

And also link this as a comment: https://github.com/aave/protocol-v2/blob/baeb455fad42d3160d571bd8d3a795948b72dd85/contracts/protocol/libraries/logic/ReserveLogic.sol#L265-L325

File: src/test/ValidateRedeemFei.sol (hunk 2 above)
Comment (0xrajath) [reply src/test/ValidateRedeemFei.sol:43]: Oh I meant to just have a check for the aFeiBalanceBefore and aFeiBalanceAfter as well. You should be able to calculate that in the test right?

**Review by 0xrajath — APPROVED** [review]
LGTM

File: src/AFeiToDaiSwapper.sol [hunk 4]
```diff
@@ -41,87 +41,72 @@ interface ILendingPool {
...
     address public constant FEI = 0x956F47F50A910163D8BF957Cf5846D573E7f87CA;
 
     address public constant A_FEI = 0x683923dB55Fead99A79Fa01A27EeC3cB19679cC3;
 
-    address public constant AAVE_MAINNET_RESERVE_FACTOR = 0x464C71f6c2F760DdA6093dCB91C24c39e5d6e18c;
-
     IFixedPricePSM public constant DAI_FIXED_PRICE_PSM = IFixedPricePSM(0x2A188F9EB761F70ECEa083bA6c2A40145078dfc2);
 
-    ILendingPool public constant AAVE_LENDING_POOL = ILendingPool(0x7d2768dE32b0b80b7a3454c06BdAc94A69DDc7A9);
+    uint256 public constant MAX_BPS = 30;
```
Comment (sakulstra) [inline src/AFeiToDaiSwapper.sol]: Perhaps 50, would be a good max as it is what was suggested on the forums for BAL acquiring.
In practice it's unlikely that the current 3bps will change at all i think, so should be good enough, no?

File: src/AFeiToDaiSwapper.sol [hunk 5]
```diff
@@ -41,87 +41,72 @@ interface ILendingPool {
...
 
-    constructor () {
-        IERC20(DAI).approve(address(AAVE_LENDING_POOL), type(uint256).max);
+    constructor() {
+        IERC20(DAI).approve(address(AaveV2Ethereum.POOL), type(uint256).max);
+        IERC20(FEI).approve(address(DAI_FIXED_PRICE_PSM), type(uint256).max);
     }
 
     function swapAllAvailable() external {
```
Comment (sakulstra) [inline src/AFeiToDaiSwapper.sol:58]: Not sure if worth it, but a little less gas intensive version would be sth like:

```
        uint256 redeemAmount = IERC20(FEI).balanceOf(address(AaveV2Ethereum.POOL));

        uint256 aFeiReserveBalance = IERC20(A_FEI).balanceOf(AaveV2Ethereum.COLLECTOR);
        if (aFeiReserveBalance < redeemAmount) {
            redeemAmount = aFeiReserveBalance;
        }

        IERC20(A_FEI).transferFrom(AaveV2Ethereum.COLLECTOR, address(this), redeemAmount);

        uint256 feiBalance = AaveV2Ethereum.POOL.withdraw(FEI, type(uint256).max, address(this));
        uint256 minBalance = feiBalance - (feiBalance * MAX_BPS / 10_000);

        // https://docs.tribedao.xyz/docs/protocol/Mechanism/PegStabilityModule
        // off by one error sometimes, so we just accept a 1 unit L
        DAI_FIXED_PRICE_PSM.redeem(address(this), feiBalance, minBalance);

        AaveV2Ethereum.POOL.deposit(DAI, IERC20(DAI).balanceOf(address(this)), AaveV2Ethereum.COLLECTOR, 0);
```

395191 vs 408242

edit: yah probably not worth it :sweat_smile: it's ~25$ and will only be called a handful of times

File: src/AFeiToDaiSwapper.sol [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+    uint256 public constant MAX_BPS = 50;
+
+    constructor() {
+        IERC20(DAI).approve(address(AaveV2Ethereum.POOL), type(uint256).max);
+        IERC20(FEI).approve(address(DAI_FIXED_PRICE_PSM), type(uint256).max);
+    }
+
+    function swapAllAvailable() external {
+        uint256 redeemAmount = IERC20(FEI).balanceOf(address(AaveV2Ethereum.POOL));
```
Comment (sakulstra) [inline src/AFeiToDaiSwapper.sol]: the important thing is the balance of aFEI, not the POOL 
```suggestion
        uint256 redeemAmount = IERC20(FEI).balanceOf(A_FEI);
```

File: src/AFeiToDaiSwapper.sol [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+            redeemAmount = aFeiReserveBalance;
+        }
+
+        IERC20(A_FEI).transferFrom(AaveV2Ethereum.COLLECTOR, address(this), redeemAmount);
+
+        AaveV2Ethereum.POOL.withdraw(FEI, redeemAmount, address(this));
+
+        uint256 feiBalance = IERC20(FEI).balanceOf(address(this));
+
+        require(DAI_FIXED_PRICE_PSM.redeemFeeBasisPoints() <= MAX_BPS);
```
Comment (sakulstra) [inline src/AFeiToDaiSwapper.sol]:
```suggestion
```
this check is redundant and should be removed

File: src/AFeiToDaiSwapper.sol [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+        uint256 minBalance = feiBalance - (feiBalance * MAX_BPS / 10_000);
+        
+        // https://docs.tribedao.xyz/docs/protocol/Mechanism/PegStabilityModule
+        uint256 outBalance = DAI_FIXED_PRICE_PSM.redeem(address(this), feiBalance, minBalance);
+        
+        require(minBalance <= outBalance);
```
Comment (sakulstra) [inline src/AFeiToDaiSwapper.sol]:
```suggestion
        require(minBalance <= outBalance, 'BALANCE_LESS_THEN_MINIMUM');
```

**Comment by sakulstra** [comment]
I think it makes sense to add a test to check that after redeem utilization is 100% and available liquidity is 0 as the current test suite didn't cover the bug here.

File: src/test/ValidateRedeemFei.sol [hunk 9]
```diff
@@ -16,31 +24,37 @@ interface IERC20 {
...
 
     function testProposalPostPayload() public {
-
-        uint256 aFeiAmount = 300_000e18;
-        uint256 daiBalanceBefore = IERC20(DAI).balanceOf(AAVE_MAINNET_RESERVE_FACTOR);
-        uint256 aFeiBalanceBefore = IERC20(A_FEI).balanceOf(AAVE_MAINNET_RESERVE_FACTOR);
+        uint256 aFeiBalance = IERC20(FEI).balanceOf(A_FEI);
+        uint256 aDaiReserveBalanceBefore = IERC20(A_DAI).balanceOf(AAVE_MAINNET_RESERVE_FACTOR);
+        uint256 aFeiReserveBalanceBefore = IERC20(A_FEI).balanceOf(AAVE_MAINNET_RESERVE_FACTOR);
+        uint256 minBalance = aFeiBalance - (aFeiBalance * 3 / 10_000);
```
Comment (sakulstra) [inline src/test/ValidateRedeemFei.sol]: I mean it's okay, but should better be MAX_BPS instead of static 3

File: src/AFeiToDaiSwapper.sol [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+    function balanceOf(address account) external view returns (uint256);
+    function approve(address spender, uint256 amount) external returns (bool);
+    function transferFrom(address from, address to, uint256 amount) external returns (bool);
+}
+
+/**
+ * @author Llama
+ * @dev This contract redeems all available aFEI in the Lending Pool for FEI, redeems FEI for DAI via Tribe DAO’s DAI Peg Stability Module (PSM), and deposits all DAI on Aave on behalf of AAVE_MAINNET_RESERVE_FACTOR.
+ * Governance Forum Post: https://governance.aave.com/t/arc-ethereum-v2-reserve-factor-afei-holding-update/9401
+ * Parameter snapshot: https://snapshot.org/#/aave.eth/proposal/0x88e896a245ffeda703e0b8f5494f3e66628be6e32a7243e3341b545c2972857f
```
Comment (sakulstra) [inline src/AFeiToDaiSwapper.sol]: this links to maticX

File: src/AFeiToDaiSwapper.sol [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+    function balanceOf(address account) external view returns (uint256);
+    function approve(address spender, uint256 amount) external returns (bool);
+    function transferFrom(address from, address to, uint256 amount) external returns (bool);
+}
+
+/**
+ * @author Llama
+ * @dev This contract redeems all available aFEI in the Lending Pool for FEI, redeems FEI for DAI via Tribe DAO’s DAI Peg Stability Module (PSM), and deposits all DAI on Aave on behalf of AAVE_MAINNET_RESERVE_FACTOR.
+ * Governance Forum Post: https://governance.aave.com/t/arc-ethereum-v2-reserve-factor-afei-holding-update/9401
+ * Parameter snapshot: https://snapshot.org/#/aave.eth/proposal/0x519f6ecb17b00eb9c2c175c586173b15cfa5199247903cda9ddab48763ddb035
```
Comment (sakulstra) [inline src/AFeiToDaiSwapper.sol]:
```suggestion
 * Snapshot: https://snapshot.org/#/aave.eth/proposal/0x519f6ecb17b00eb9c2c175c586173b15cfa5199247903cda9ddab48763ddb035
```

Nitpick, but not really a parameter check right?

**Review by sakulstra — APPROVED** [review]
